File: src/python/xerxes/runtime/transcript.py
# Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""In-memory transcript store backing every :class:`QueryEngine` session.

:class:`TranscriptEntry` holds one role/content/metadata triple; :class:`TranscriptStore`
keeps the ordered list that is serialised in ``QueryEngine.to_dict`` and
converted to provider message dicts via
:meth:`TranscriptStore.to_messages`.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime
from typing import Any


@dataclass
class TranscriptEntry:
    """One message in the transcript.

    Attributes:
        role: Conversational role (``"user"``, ``"assistant"``, ``"system"``,
            ``"tool"``).
        content: Rendered text content of the message.
        timestamp: ISO-8601 timestamp captured when the entry was constructed.
        metadata: Provider-specific extras (tool call ids, attachments, etc.)
            merged into the message dict by :meth:`TranscriptStore.to_messages`.
    """

    role: str
    content: str
    timestamp: str = field(default_factory=lambda: datetime.now().isoformat())
    metadata: dict[str, Any] = field(default_factory=dict)


@dataclass
class TranscriptStore:
    """Ordered list of :class:`TranscriptEntry` records with compaction helpers.

    Attributes:
        entries: All messages in chronological order.
        flushed: Whether the store has been persisted since its last write.
    """

    entries: list[TranscriptEntry] = field(default_factory=list)
    flushed: bool = False

    def append(self, role: str, content: str, **metadata: Any) -> None:
        """Append a new entry built from ``role``, ``content`` and ``metadata``."""
        self.entries.append(TranscriptEntry(role=role, content=content, metadata=metadata))
        self.flushed = False

    def append_entry(self, entry: TranscriptEntry) -> None:
        """Append an already-constructed :class:`TranscriptEntry`."""
        self.entries.append(entry)
        self.flushed = False

    def replay(self) -> tuple[TranscriptEntry, ...]:
        """Return an immutable snapshot of every stored entry, in order."""
        return tuple(self.entries)

    def to_messages(self) -> list[dict[str, Any]]:
        """Render entries as provider-style ``{"role", "content", ...}`` dicts.

        Metadata fields are merged onto each dict so tool-call ids and other
        provider-specific keys round-trip back into the streaming loop.
        """

        messages = []
        for entry in self.entries:
            msg: dict[str, Any] = {"role": entry.role, "content": entry.content}
            msg.update(entry.metadata)
            messages.append(msg)
        return messages

    def flush(self) -> None:
        """Mark the in-memory store as persisted by the caller."""
        self.flushed = True

    def clear(self) -> None:
        """Drop every entry and reset the flushed flag."""
        self.entries.clear()
        self.flushed = False

    @property
    def turn_count(self) -> int:
        """Number of user turns (entries with ``role == "user"``)."""
        return sum(1 for e in self.entries if e.role == "user")

    @property
    def message_count(self) -> int:
        """Total stored entries regardless of role."""
        return len(self.entries)

    def as_markdown(self) -> str:
        """Render the transcript as a truncated Markdown bullet list."""

        lines = ["# Transcript", "", f"Messages: {self.message_count}", ""]
        for entry in self.entries:
            role_tag = f"**{entry.role}**"
            content_preview = entry.content[:200]
            if len(entry.content) > 200:
                content_preview += "..."
            lines.append(f"- {role_tag}: {content_preview}")
        return "\n".join(lines)


__all__ = [
    "TranscriptEntry",
    "TranscriptStore",
]

File: src/typescript/src/runtime/transcript.ts
// Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
// Licensed under the Apache License, Version 2.0.

export interface TranscriptEntryOptions {
  readonly content: string
  readonly metadata?: Readonly<Record<string, unknown>>
  readonly role: string
  readonly timestamp?: string
}

/** Immutable message record retained by the in-memory transcript store. */
export class TranscriptEntry {
  readonly content: string
  readonly metadata: Readonly<Record<string, unknown>>
  readonly role: string
  readonly timestamp: string

  constructor(options: TranscriptEntryOptions) {
    this.role = textValue(options.role, 'role')
    this.content = textValue(options.content, 'content')
    this.timestamp = timestampValue(options.timestamp ?? new Date().toISOString())
    this.metadata = freezeRecord(options.metadata ?? {})
    Object.freeze(this)
  }
}

export interface TranscriptStoreOptions {
  readonly now?: () => Date
}

/** Ordered transcript with replay, provider-message, and compact display helpers. */
export class TranscriptStore {
  private readonly clock: () => Date
  private readonly ledger: TranscriptEntry[] = []
  private persisted = false

  constructor(options: TranscriptStoreOptions = {}) {
    this.clock = options.now ?? (() => new Date())
  }

  get entries(): readonly TranscriptEntry[] {
    return this.ledger.map(copyEntry)
  }

  get flushed(): boolean {
    return this.persisted
  }

  get turnCount(): number {
    return this.ledger.filter(entry => entry.role === 'user').length
  }

  get messageCount(): number {
    return this.ledger.length
  }

  append(role: string, content: string, metadata: Readonly<Record<string, unknown>> = {}): void {
    this.ledger.push(new TranscriptEntry({
      role,
      content,
      metadata,
      timestamp: this.nowTimestamp(),
    }))
    this.persisted = false
  }

  appendEntry(entry: TranscriptEntry): void {
    if (!(entry instanceof TranscriptEntry)) throw new TypeError('entry must be a TranscriptEntry')
    this.ledger.push(copyEntry(entry))
    this.persisted = false
  }

  replay(): readonly TranscriptEntry[] {
    return Object.freeze(this.ledger.map(copyEntry))
  }

  /** Merge provider-specific metadata onto each basic role/content message. */
  toMessages(): ReadonlyArray<Readonly<Record<string, unknown>>> {
    return this.ledger.map(entry => Object.freeze({
      role: entry.role,
      content: entry.content,
      ...cloneRecord(entry.metadata),
    }))
  }

  flush(): void {
    this.persisted = true
  }

  clear(): void {
    this.ledger.length = 0
    this.persisted = false
  }

  asMarkdown(): string {
    const lines = ['# Transcript', '', 'Messages: ' + this.messageCount, '']
    for (const entry of this.ledger) {
      const preview = entry.content.length > 200 ? entry.content.slice(0, 200) + '...' : entry.content
      lines.push('- **' + entry.role + '**: ' + preview)
    }
    return lines.join('\n')
  }

  private nowTimestamp(): string {
    const now = this.clock()
    if (!(now instanceof Date) || Number.isNaN(now.valueOf())) throw new RangeError('now must return a valid Date')
    return now.toISOString()
  }
}

function copyEntry(entry: TranscriptEntry): TranscriptEntry {
  return new TranscriptEntry({
    role: entry.role,
    content: entry.content,
    timestamp: entry.timestamp,
    metadata: entry.metadata,
  })
}

function textValue(value: unknown, name: string): string {
  if (typeof value !== 'string') throw new TypeError(name + ' must be a string')
  return value
}

function timestampValue(value: unknown): string {
  if (typeof value !== 'string' || Number.isNaN(new Date(value).valueOf())) {
    throw new RangeError('timestamp must be a valid ISO timestamp')
  }
  return value
}

function cloneRecord(record: Readonly<Record<string, unknown>>): Readonly<Record<string, unknown>> {
  return freezeRecord(record)
}

function freezeRecord(record: Readonly<Record<string, unknown>>): Readonly<Record<string, unknown>> {
  const copy: Record<string, unknown> = {}
  for (const [key, value] of Object.entries(record)) copy[key] = freezeValue(value)
  return Object.freeze(copy)
}

function freezeValue(value: unknown): unknown {
  if (Array.isArray(value)) return Object.freeze(value.map(freezeValue))
  if (value instanceof Date) return value.toISOString()
  if (value && typeof value === 'object') return freezeRecord(value as Record<string, unknown>)
  return value
}
